metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala

```diff
@@ -0,0 +1,389 @@
+      val thread = new Thread(s"Evaluating Worksheet ${path.filename}") {
+        override def run(): Unit = {
+          result.complete(evaluateWorksheet(path, token))
+        }
+      }
+      cancelables.add(
+        Cancelable(() => {
+          if (thread.isAlive) {
+            thread.stop()
+          }
+        })
+      )
+      stopThreadOnCancel(path, result, thread)
+      thread.start()
+      thread.join()
+    }
+    jobs.submit(
+      result,
+      () => {
+        try runEvaluation()
+        catch onError
+      }
+    )
+    result.asScala.recover(onError)
+  }
+
+  /**
+   * Prompts the user to cancel the task after a few seconds.
+   *
+   * Attempts to gracefully shut down the thread when users requests to cancel:
+   * First tries `Thread.interrupt()` with fallback to `Thread.stop()` after
+   * one second if interruption doesn't work.
+   */
+  private def stopThreadOnCancel(
+      path: AbsolutePath,
+      result: CompletableFuture[Array[DecorationOptions]],
+      thread: Thread
+  ): Unit = {
+    // Last resort, if everything else fails we use `Thread.stop()`.
+    val stopThread = new Runnable {
+      def run(): Unit = {
+        if (thread.isAlive()) {
+          scribe.warn(s"thread stop: ${thread.getName()}")
+          Cancelable.cancelAll(
+            List(
+              Cancelable(() => thread.stop()),
+              cancelables
+            )
+          )
+        }
+      }
+    }
+    // If the program is running for more than
+    // `userConfig().worksheetCancelTimeout`, then display a prompt for the user
+    // to cancel the program.
+    val promptUserToCancel = new Runnable {
+      def run(): Unit = {
+        if (!result.isDone()) {
+          val cancel = languageClient.metalsSlowTask(
+            new MetalsSlowTaskParams(
+              s"Evaluating worksheet '${path.filename}'",
+              noLogs = true,
+              secondsElapsed = userConfig().worksheetCancelTimeout
+            )
+          )
+          cancel.asScala.foreach { c =>
+            if (c.cancel && thread.isAlive()) {
+              // User has requested to cancel a running program. first line of
+              // defense is `Thread.interrupt()`. Fingers crossed it's enough.
+              result.complete(Array.empty)
+              threadStopper.schedule(stopThread, 1, TimeUnit.SECONDS)
+              scribe.warn(s"thread interrupt: ${thread.getName()}")
+              thread.interrupt()
+            }
+          }
+          result.asScala.onComplete(_ => cancel.cancel(true))
+        }
+      }
+    }
+    threadStopper.schedule(
+      promptUserToCancel,
+      userConfig().worksheetCancelTimeout,
+      TimeUnit.SECONDS
+    )
+  }
+
+  private def renderHoverMessage(
+      statement: Statement,
+      margin: Int,
+      isEmptyValue: Boolean
+  ): String = {
+    val out = new StringBuilder()
+    if (!isEmptyValue) {
+      statement.binders.iterator.foreach { binder =>
+        out
```

Comment:
Maybe this would be a bit more clear:
```
val color = binder.tpe.render(TPrintColors.BlackWhite)
out.append(s"\n${binder.name}: $color: ")
``` 
?

Reply:
Not really IMO. When using string builders it's usually best to append individual strings